### ISSUE TITLE
Accelerator#_prepare_fsdp turns AdamW 8-bit optimizer into AdamW 32-bit

### ISSUE THREAD

**Issue by scottlogic-alex** [huggingface/accelerate#1902]
### System Info

```Shell
- `Accelerate` version: 0.21.0
- Platform: Linux-5.19.0-45-generic-x86_64-with-glibc2.35
- Python version: 3.11.2
- Numpy version: 1.25.2
- PyTorch version (GPU?): 2.1.0.dev20230802+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 62.79 GB
- GPU type: NVIDIA A40
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: FSDP
        - mixed_precision: bf16
        - use_cpu: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - fsdp_config: {'fsdp_auto_wrap_policy': 'TRANSFORMER_BASED_WRAP', 'fsdp_backward_prefetch_policy': 'NO_PREFETCH', 'fsdp_forward_prefetch': False, 'fsdp_offload_params': False, 'fsdp_sharding_strategy': 1, 'fsdp_state_dict_type': 'SHARDED_STATE_DICT', 'fsdp_sync_module_states': True, 'fsdp_transformer_layer_cls_to_wrap': 'LlamaDecoderLayer', 'fsdp_use_orig_params': True}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Training script is a [fork of `qlora.py`](https://github.com/scottlogic-alex/qlora/blob/d5e564ad3affcaf62c22912ba36cab2d77c2d585/qlora.py).

But if you have a preferred script, go ahead and use that; the triggers are:

- FSDP enabled
- AdamW 8-bit optimizer selected

### Configure accelerate

Set up your accelerate config (`~/.cache/huggingface/accelerate/default_config.yaml`) like so:

```yaml
compute_environment: LOCAL_MACHINE
distributed_type: FSDP
downcast_bf16: 'no'
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch_policy: NO_PREFETCH
  fsdp_forward_prefetch: false
  fsdp_offload_params: false
  fsdp_sharding_strategy: 1
  fsdp_state_dict_type: SHARDED_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_use_orig_params: true
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

### Set up repository

```bash
git clone https://github.com/scottlogic-alex/qlora.git
cd qlora
git checkout d5e564ad3affcaf62c22912ba36cab2d77c2d585
pip install -U -r requirements.txt
```

### Run trainer

We will run `qlora.py` on 2xA40, non-distributed (it's one coreweave computer with 2 GPUs).  
This config will atempt to run a full finetune of Llama 2 7b.

```bash
python -m accelerate.commands.launch qlora.py --use_auth_token True --model_name_or_path meta-llama/Llama-2-7b-chat-hf --trust_remote_code --dataset prm800k-solutions --dataset_format prm800k-solutions --max_memory_MB 48000 --simulate_worst_case_seq_len --truncate_toward_center --source_max_len 2048 --target_max_len 2048 --gradient_accumulation_steps 2 --per_device_train_batch_size 1 --per_device_eval_batch_size 1 --optim adamw_8bit --learning_rate 0.0002 --save_steps 196 --save_total_limit 1 --max_steps 256 --evaluation_strategy steps --eval_steps 196 --measure_memory --terminate_after_step 16 --bits 32 --full_finetune 
```

### Verify that optimizer begins as 8-bit

I placed a breakpoint at the end of `transformers.trainer::Trainer#create_optimizer`:  
https://github.com/huggingface/transformers/blob/4c21da5e347bfc53ee4ec5b71a23721fefe6822c/src/transformers/trainer.py#L1001

`self.optimizer.args.optim_bits == 8`, as expected of AdamW 8-bit:

<img width="962" alt="image" src="https://github.com/huggingface/accelerate/assets/40387940/58ba0c81-ce36-4c07-9a24-a6ffb9bf3beb">

### Verify that `accelerate` recreates our optimizer as AdamW 32-bit

I placed a breakpoint in `accelerate.accelerator::Accelerator#_prepare_fsdp`:  
https://github.com/huggingface/accelerate/blob/bc86a9379f555f67504b9877171c66e4c588bd14/src/accelerate/accelerator.py#L1114

Observe how the optimizer that's submitted to `_prepare_fsdp` is the same 8-bit AdamW that we saw earlier:  
<img width="871" alt="image" src="https://github.com/huggingface/accelerate/assets/40387940/1a1d36d7-5f40-47e6-9954-74ed3717c025">

_Side-note: construction fails, due to `differentiable` arg present in `obj.optimizer.defaults` (https://github.com/huggingface/accelerate/issues/801), so it falls-back to removing the `differentiable` arg and trying again:_
<img width="870" alt="image" src="https://github.com/huggingface/accelerate/assets/40387940/9ad1db47-7231-4029-bfab-b69cab61a333">

After `accelerate` creates our new optimizer, observe how it's **not the same as before**; `obj.optimizer.args.optim_bits != optimizer.args.optim_bits`:  
<img width="866" alt="image" src="https://github.com/huggingface/accelerate/assets/40387940/395398df-3786-4abe-88cc-ce01fd8e7a39">

### Run out of memory

Since our 8-bit optimizer got turned into a 32-bit optimizer: we don't have enough memory to full-finetune Llama 7b in mixed-precision on 2xA40. 😢 

<img width="1477" alt="image" src="https://github.com/huggingface/accelerate/assets/40387940/736a8b00-699e-4436-bae8-5138ff28ad35">

### Expected behavior

When accelerate prepares our optimizer for FSDP: 8-bit optimizers should **remain** 8-bit after preparation. 🙃

**Comment by pacman100** [comment]
Hello, could you try with the latest release of Transformers and Accelerate as it was fixed in that?

**Comment by scottlogic-alex** [comment]
awesome! yes, confirmed fixed after I upgraded accelerate 0.21.0 -> 0.22.0 and transformers 4.31.0->4.32.1.

<img width="912" alt="image" src="https://github.com/huggingface/accelerate/assets/40387940/faf0f5a9-7355-4dee-9011-25ffd3585f6d">

debugged using:

```python
assert next(iter(self.optimizer.state.items()))[0] is model.get_input_embeddings().weight
(lambda param: print(f"""
optimizer param 0 (input embedding)
reference to flattened params (model.get_input_embeddings().weight): {param[0].dtype} {param[0].shape} {param[0].device}
momentum: {param[1]['state1'].dtype} {param[1]['state1'].shape} {param[1]['state1'].device}
variance: {param[1]['state2'].dtype} {param[1]['state2'].shape} {param[1]['state2'].device}
qmap1: {param[1]['qmap1'].dtype} {param[1]['qmap1'].shape} {param[1]['qmap1'].device}
qmap2: {param[1]['qmap2'].dtype} {param[1]['qmap2'].shape} {param[1]['qmap2'].device}
absmax1: {param[1]['absmax1'].dtype} {param[1]['absmax1'].shape} {param[1]['absmax1'].device}
absmax2: {param[1]['absmax2'].dtype} {param[1]['absmax2'].shape} {param[1]['absmax2'].device}
"""))(next(iter(self.optimizer.state.items())))
```

at the `on_step_end` callback in `transformers.trainer::Trainer#_inner_training_loop`:  
https://github.com/huggingface/transformers/blob/dc0c102954ff1f6bcb47de85afea5edc81fc8c7f/src/transformers/trainer.py#L1927

FSDP mixed-precision training of llama 2 7b on 2xA40 **no longer OOMs** 🎉 .

memory usage / iteration speed:

FSDP with flash attention 2:  
step 1 95993MiB / 98280MiB  
step 2 81011MiB / 98280MiB  
step 3 80993MiB / 98280MiB (memory stays at this level 'til step 11, where I terminated it)  
9.88s/it

FSDP without flash attention 2:  
step 1 94065MiB / 98280MiB  
step 2 84727MiB / 98280MiB  
step 3 71849MiB / 98280MiB
step 4 75757MiB / 98280MiB
step 5 77919MiB / 98280MiB
step 6 78641MiB / 98280MiB
21~27s/it

but using vertical sharding (`device_map='auto') instead of horizontal sharding (FSDP) ran faster, which is a bit unexpected (because vertical sharding means leaving most GPUs idle most of the time). both these A40 GPUs are in the same coreweave machine, but I don't know how fast the interconnect is for them to share weights via FSDP.

`device_map='auto' with flash attention 2:  
step 1 74061MiB / 98280MiB
step 2 85107MiB / 98280MiB
step 3 85607MiB / 98280MiB (memory stays at this level 'til step 15, where I terminated it)  
5.77s/it

`device_map='auto' without flash attention 2:  
step 1 88365MiB / 98280MiB
step 2 OOM
